Add support for grain state classes

sergeybykov opened the pull request:
This is to allow passing a concrete grain state class as a type argument to Grain<T>. This was initially driven by the need to have a way to define custom logic of applying events to state for Event Sourcing - https://github.com/dotnet/orleans/issues/343. In addition this would allow us later to get rid of the codegen of state classes altogether.

This is a breaking change: `Read/Write/ClearStateAsync` methods got moved from `IGrainState` to `Grain<T>`, so that state classes don't have to implement them. Very easy to adjust existing code by simply removing `State.` in those statements. We may consider changing the methods to take a state object as an argument, but that's a separate decision to make, and it's not part of this change. This is only what @yevhen called "separating *unit-of-work* from *state*." Consider this just a first step in refactoring the state management story and adding support for Event Sourcing.

nehmebilal commented on src/Orleans/Core/Grain.cs:
```diff
@@ -272,9 +275,25 @@ internal string CaptureRuntimeEnvironment()
     {
+        private IStorage stateStorageBridge;
+
+        private IStorage Storage
```
This breaks DI and unit testing of grains. `IStorage` should be injected through the constructor (potentially be made part of the `IGrainRuntime`). Which implementation of `IStorage` to use should be decided on the top level (Silo initialization) and passed in.

sergeybykov replied:
Good point. I missed that. I'll try to refactor it.

sergeybykov replied:
@nehmebilal Done. Check it out.

nehmebilal commented on src/Orleans/Runtime/GrainRuntime.cs:
```diff
@@ -8,18 +8,15 @@ namespace Orleans.Runtime
-        private readonly IStreamProviderManager streamProviderManager;
 
-        public GrainRuntime(string id, IGrainFactory grainFactory, ITimerRegistry timerRegistry, IReminderRegistry reminderRegistry, IStreamProviderManager streamProviderManager)
+        public GrainRuntime(string id, IGrainFactory grainFactory, ITimerRegistry timerRegistry, IReminderRegistry reminderRegistry, IStreamProviderManager streamProviderManager, IStorage storage=null)
```
Why storage can be null? This is relying on a logic of assigning a default storage in the `Grain`. This hurts code readability as it's impossible for someone new looking at this class to know what's going on.

nehmebilal commented on src/Orleans/Core/Grain.cs:
```diff
@@ -272,9 +275,25 @@ internal string CaptureRuntimeEnvironment()
+        {
+            get
+            {
+                if (storage == null) // Of storage has been injected via DI, we'll use it. Otherwise, we initialize here.
```
I understand that this makes it easier but it looks more like a work around to avoid plumbing the storage. Instantiating the `IStorage` implementation should not be done in the `Grain`. I would recommend plumbing the `IStorage` the same way we did with other IGrainRuntime components as the current approach can have side effects.

sergeybykov commented:
>I understand that this makes it easier but it looks more like a work around to avoid plumbing the storage. Instantiating the  IStorage  implementation should not be done in the  Grain . I would recommend plumbing the  IStorage  the same way we did with other IGrainRuntime components as the current approach can have side effects.

@nehmebilal I'll let @jthelin chime in but my interpretation is his intent here was to delay instantiation of the bridge object until it's needed, with a chance that read-only requests would not trigger instantiation of it at all.

I'll play around a bit more to see how we can make it cleaner.

nehmebilal commented:
There is another issue: There is on `IStorage` per-grain and not per-runtime. So adding `IStorage` to `IGrainRuntime` doesn't work. It's the `IStorageProvider` that needs to be added to the `IGrainRuntime` and used to get access to storage. 

To keep the lazy create, you can add a **factory method** to `IStorageProvider`, something like:
```csharp
IStorage CreateStorage(string grainTypeName, Grain<TGrainState> grain, IStorageProvider store);
```

Then inside the `Grain`:
```csharp
        private IStorage Storage
        {
            get
            {
                if (storage == null) // Of storage has been injected via DI, we'll use it. Otherwise, we initialize here.
                {
                    // Lazy create the storage bridge
                    storage = Runtime.StorageProvider.CreateStorage();
                }
                return storage;
            }
        }
```

sergeybykov commented:
Yep, I reached the same conclusion that adding Storage to IGrainRuntime was a mistake. I'm looking if I can make Catalog set Storage property instead.

nehmebilal commented:
Yes but you'll still have the testing issue (calls to `IStorage` cannot be stubbed) unless you add `IStorage` to the constructor of `Grain` or you use the factory method suggested above and make the `IStorageProvider` part of `IGrainRuntime`. I would say that `IStorageProvider` should be part of `IGrainRuntime` in any case (and if possible, it would be great if we can remove `IStorageProvider` from `IActivationData`).

sergeybykov commented:
Yes, I was thinking about passing `IStorage` to grain's constructor. I'm less sure about exposing IStorageProvider directly to grain. We may need a layer of indirection there.

Maybe we should start with passing/setting IStorage for now as a first step. The bridge is a temporary solution anyway. Plus we are likely to have a different storage provider interface for Event Sourcing grains.

nehmebilal commented:
> Maybe we should start with passing/setting IStorage for now as a first step.

Sounds good! that'll make testing even easier and the `Grain` class simpler.

Btw you can even use a **decorator** for `IStorage` that does the lazy create (if we really need the lazy creation). Something like:

```csharp
public class LazyStorage<TGrainState> : IStorage
{
	private IStorage delegate;
	
	private readonly IStorageProvider store;
    private readonly Grain<TGrainState> grain;
    private readonly string grainTypeName;	
	
	public LazyStorage(string grainTypeName, Grain<TGrainState> grain, IStorageProvider store)
	{
        this.grainTypeName = grainTypeName;
        this.grain = grain;
        this.store = store;
	}
	
	public async Task ReadStateAsync()
	{
		LazyInit();
		delegate.ReadStateAsync();
	}

	public async Task WriteStateAsync()
	{
		LazyInit();
		delegate.WriteStateAsync();
	}

	private void LazyInit()
	{
		if (delegate == null)
		{
			// it's better to use an abstract factory here
			delegate = new GrainStateStorageBridge<TGrainState>(grainTypeName, this, store);
		}
	}
}
```

jthelin commented:
`GrainStateStorageBridge` was originally written before all the recent DI / grain testability changes, so some tweaks are quite likely to be required to reflect that new thinking. :)

Having said that, given that the **only** thing that `GrainStateStorageBridge` does is to connect a `GrainState` object to a `StorageProvider` object, then I don 't see why we would need / want to inject all three.

If the DI principle for constructor parameters is "tell me what hookups I need to make this class work", then isn't injecting EITHER `GrainState` and `StorageProvider` OR `GrainState` and `Storage` exactly the semantics we want to convey?

It does seem like the most natural and "grokkable" mapping would be to inject `State` and `Storage` [because that is what the grain code interacts with directly.]

I feel we can maybe just forget about lazy-create now if we inject the Storage property directly. 
For runtime, then if a grain was declared as stateful grain then the runtime will always explicitly create an `IStorage` "delegate bridge" to bind `State` and `Storage` together and if grain has no state then runtime will inject `null` for both?

So seems like the plan is converging at inject `IStorage` constructor parameter into grain?

nehmebilal commented:
> It does seem like the most natural and "grokkable" mapping would be to inject State and Storage [because that is what the grain code interacts with directly.]

Yes, that makes a lot of sense.

On the other hand, when we look at the `Grain` constructor, we see *Identification*, *State* and *Runtime*; and it feels like the interaction with the storage is part of the interaction with the runtime (i.e. belongs in `IGrainRuntime`). 

If we have a reason to treat the interaction with storage differently than other Runtime-Grain interaction, then injecting the storage directly is the way to go. Otherwise, making `IStorageProvider` part of `IGrainRuntime` would be something to consider. After all, as you mentioned, the `GrainStateStorageBridge` is not doing more than delegating to the `IStorageProvider` and doesn't necessarily need to be injected. I would personally remove this level of indirection and make the `Grain` talk directly to the `IStorageProvider` (via `IGrainRuntime`).

Either way, the code will be testable :)

jthelin commented:
We ended up with choosing a good name for `IGrainRuntime` because it is really about "My Runtime" from the grains perspective -- i.e. a view of the "local system" runtime tailored to that grain -- for example "my" deactivation control and "my" reminders and "my" timers for this grain.

We also have slightly broken that "my runtime" abstraction by including some "global" accessor functions into `IGrainRuntime` for grain factories and stream providers. 
Although in theory the underlying implementation could easily be tweaked later on if desired without affecting the API so that `IGrainRuntime` returns just a subset of those "global" system views tailored to "my grain". 
For example, I can imagine that we later on might want to apply some security capability / scoping mechanism so that not all grains were necessarily able to access "all" stream providers or grain, just the ones they were pre-authorized to use.

In the case of `Storage` we are talking about "my" storage provider instance for this grain, so including that in `IGrainRuntime` definitely matches the "my runtime" abstraction.

If I can re-cast my vote, then I do have a slight preference based on an API abstraction perspective for mapping `Storage` in via the "my runtime" route. 
The DI Purists might disagree, but I think we have a good API abstraction around injecting "my runtime" as a unit, and it also avoids the extra code sprawl of a long list of grain constructor params :)

But I am fine either way if we decide to inject `Storage` via "my runtime" or as a first-class constructor param. 

Here, let me fix that sentence for you :)
> Either way, the code will be [unit] testable :)

nehmebilal commented:
> We ended up with choosing a good name for IGrainRuntime because it is really about "My Runtime" from the grains perspective -- i.e. a view of the "local system" runtime tailored to that grain -- for example "my" deactivation control and "my" reminders and "my" timers for this grain.

I agree but it's not exactly the case right now :) When the grain talks to `ITimerRegistry` or `IReminderRegistry` it passes `this` in the call. We have hidden direct access to other grains timers and reminders by making the `ITimerRegistry` and `IReminderRegistry` `internal` so `Grain` implementers can't call it directly (except via exposed `protected` methods). It just feels inconsistent to treat storage differently.

Oh and now that I think about it, there will be a circular dependency because the `IGrainRuntime` will depend on the `Grain` while the `Grain` depends on `IGrainRuntime`.

> Here, let me fix that sentence for you :)
Either way, the code will be [unit] testable :)

:8ball:

nehmebilal commented on src/Orleans/Core/Grain.cs:
```diff
@@ -308,5 +312,20 @@ protected TGrainState State
             get { return base.GrainState as TGrainState; }
         }
+
+        public Task ClearStateAsync()
```
Should these be `protected`?

sergeybykov replied:
It's not there anymore.

nehmebilal replied:
How? I can still see it. Do you want to allow a `Grain` to clear/read/write another `Grain`'s state?

sergeybykov replied:
Of course you are right. :-) Sorry, I looked at a wrong branch. Currently they need to be public for `GrainState.Read/Write/ClearStateAsync` methods marked obsolete to work. I'll try to rewire those via `Grain.Storage` instead.

sergeybykov replied:
Done with an additional simplification and a added test for the grain interface case.

jthelin commented on src/Orleans/CodeGeneration/GrainState.cs:
```diff
@@ -112,175 +113,107 @@ protected GrainState(string grainTypeFullName)
+        /// <summary>
+        /// Resets properties of the state object to their default values.
+        /// </summary>
+        private void ReesetProperties()
```
Typo in name.

sergeybykov replied:
Fixed.

jthelin commented on src/Orleans/CodeGeneration/GrainState.cs:
```diff
@@ -112,175 +113,107 @@ protected GrainState(string grainTypeFullName)
         /// Async method to cause refresh of the current grain state data from backing store.
         /// Any previous contents of the grain state data will be overwritten.
         /// </summary>
+        [Obsolete("Use Grain.Storage.ReadStateAsync method instead.")]
```
Shouldn't this now be "Use Grain.ReadStateAsync method instead." ?

sergeybykov replied:
Fixed.

jthelin commented:
LGTM, modulo a few minor typos.

sergeybykov commented:
I pushed one more commit - to add explicit checks for not-specified StorageProvider attribute for grains extending Grain<T> and another one for messages in [Obsolete].

sergeybykov commented:
I hope this is now ready to be rebased/squashed and merged.

sergeybykov commented:
I rebased and squashed some commits in https://github.com/sergeybykov/orleans/tree/state-classes-5. I can close this PR and submit https://github.com/sergeybykov/orleans/tree/state-classes-5 for merging.

sergeybykov commented:
I submitted a rebased and partly squashed version of this change as https://github.com/dotnet/orleans/pull/476. Closing this one.
